File: cli/test-tui.py
#!/usr/bin/env python3
"""
TUI observability harness for aristotle.

Spawns `node bin/aristotle.js <topic>` under a real PTY (via Python's built-in
`pty` module — no native deps) so Ink renders like a real terminal. Captures
every stdout chunk with a monotonic timestamp, scripts turn 2 (diagnosis
answer) and turn 3 (approval), and scans for static windows (gaps > threshold
with no screen updates).

Why this matters: Ink redraws its live area on every React state change. A
spinner/progress bar that's animating triggers a redraw every 80-400ms, so
any gap > ~1s in PTY output means nothing was animating — the display was
static, which is the "50 seconds of nothing" bug.

Usage:
    python3 test-tui.py "coin flips"
    python3 test-tui.py "coin flips" --max-gap 2.0
"""
import pty, os, sys, time, json, select, re, signal, argparse

ARISTOTLE_BIN = os.path.join(os.path.dirname(os.path.abspath(__file__)), "bin", "aristotle.js")
ARGS = argparse.ArgumentParser()
ARGS.add_argument("topic", nargs="?", default="coin flips")
ARGS.add_argument("--max-gap", type=float, default=2.0, help="fail if any PTY output gap exceeds this (seconds)")
ARGS.add_argument("--hard-timeout", type=float, default=900.0, help="overall hard timeout (seconds)")
ARGS.add_argument("--log-dir", default="/tmp")
ns = ARGS.parse_args()

stamp = int(time.time() * 1000)
raw_log = os.path.join(ns.log_dir, f"aristotle-tui-{stamp}.log")
json_log = raw_log + ".chunks.jsonl"
with open(raw_log, "w") as f: pass
with open(json_log, "w") as f: pass

started = time.monotonic()
def ts(): return f"{time.monotonic() - started:7.2f}s"
def log(*a): print(f"[{ts()}]", *a, file=sys.stderr, flush=True)

log(f"topic     : {ns.topic}")
log(f"raw log   : {raw_log}")
log(f"json log  : {json_log}")
log(f"max gap   : {ns.max_gap}s")

# --- Fork a PTY ---
pid, fd = pty.fork()
if pid == 0:
    # child: exec node bin/aristotle.js <topic>
    os.execvp("node", ["node", ARISTOTLE_BIN, ns.topic])

# parent:
chunks = []
buffer = b""
last_data_at = time.monotonic()
stage = 0  # 0=await-diagnosis, 1=await-outline, 2=await-done
staged_at = {0: time.monotonic()}
saw_done = False

ANSI_RE = re.compile(rb"\x1b\[[0-9;?]*[A-Za-z]|\x1b[()][0-9A-Za-z]")

def stripped(b):
    return ANSI_RE.sub(b"", b).decode("utf-8", errors="replace")

def write_chunk(data, gap_s):
    now = time.monotonic() - started
    entry = {
        "t": round(now, 3),
        "gap": round(gap_s, 3),
        "bytes": len(data),
        "text": stripped(data)[-200:],
    }
    chunks.append(entry)
    with open(json_log, "a") as f:
        f.write(json.dumps(entry) + "\n")
    with open(raw_log, "ab") as f:
        f.write(data)

def send(text, label):
    log(f"→ send [{label}]: {text!r}")
    os.write(fd, (text + "\r").encode("utf-8"))

def visible_text():
    return stripped(buffer)

hard_deadline = time.monotonic() + ns.hard_timeout

try:
    while True:
        if time.monotonic() > hard_deadline:
            log("hard timeout reached, killing")
            break
        r, _, _ = select.select([fd], [], [], 0.5)
        if r:
            try:
                data = os.read(fd, 8192)
            except OSError:
                break
            if not data:
                break
            now = time.monotonic()
            gap = now - last_data_at
            write_chunk(data, gap)
            buffer += data
            last_data_at = now

            txt = visible_text()
            if stage == 0 and re.search(r"question|diagnos|know|calibrat", txt, re.I) and time.monotonic() - staged_at[0] > 3:
                time.sleep(1.0)
                send("I don't know any of the answers to your questions.", "turn 2 (ignorance)")
                stage = 1; staged_at[1] = time.monotonic()
            elif stage == 1 and re.search(r"approve|sound good|sound right|proceed|want me to|ready to write", txt, re.I) and time.monotonic() - staged_at[1] > 3:
                time.sleep(1.0)
                send("ok", "turn 3 (approval)")
                stage = 2; staged_at[2] = time.monotonic()
            elif stage == 2 and re.search(r"open .*\.html|breakdown\.html", txt, re.I):
                saw_done = True
                log("saw completion hint")
                # give it a moment to exit cleanly
                try:
                    os.waitpid(pid, os.WNOHANG)
                except ChildProcessError:
                    pass
                break
        else:
            # no data for 500ms; check whether child exited
            try:
                wpid, status = os.waitpid(pid, os.WNOHANG)
                if wpid != 0:
                    log(f"child exited (status={status})")
                    break
            except ChildProcessError:
                break
finally:
    try:
        os.kill(pid, signal.SIGTERM)
        time.sleep(0.3)
        os.kill(pid, signal.SIGKILL)
    except (ProcessLookupError, PermissionError):
        pass
    try: os.close(fd)
    except: pass

# --- Analyze ---
log("\n— analysis —")
log(f"chunks captured   : {len(chunks)}")
if not chunks:
    log("  🔴 no output captured at all")
    sys.exit(3)
log(f"total duration    : {chunks[-1]['t']:.1f}s")

gaps = sorted([c for c in chunks if c["gap"] > 1.0], key=lambda c: -c["gap"])
log(f"gaps > 1.0s       : {len(gaps)}")
for g in gaps[:15]:
    preview = g["text"].strip().replace("\n", " ⏎ ")[:100]
    log(f"  gap={g['gap']:5.2f}s  @ t={g['t']:6.1f}s  next=\"{preview}\"")

longest = max(chunks, key=lambda c: c["gap"])
log(f"longest gap       : {longest['gap']:.2f}s @ t={longest['t']:.1f}s")
log(f"saw completion    : {saw_done}")

fails = [g for g in gaps if g["gap"] > ns.max_gap]
if fails:
    log(f"\n  🔴 FAIL — {len(fails)} static windows exceeded {ns.max_gap}s (longest {longest['gap']:.2f}s)")
    log(f"  full chunk log: {json_log}")
    sys.exit(2)
else:
    log(f"\n  ✅ PASS — no static window exceeded {ns.max_gap}s (longest {longest['gap']:.2f}s)")
    sys.exit(0)

File: cli/test-tui.js
/**
 * TUI observability test — actually runs aristotle under a PTY, captures
 * every stdout chunk with a timestamp, and scans for static windows.
 *
 * Ink redraws its live area on every state change (typically multiple times
 * per second whenever a spinner or progress bar is active). So any gap >1s
 * in stdout output means NOTHING was animating — the display was static.
 * That's the "50 seconds of nothing" bug the user sees.
 */
import pty from 'node-pty';
import { resolve, dirname } from 'path';
import { fileURLToPath } from 'url';
import { writeFileSync, appendFileSync, rmSync, existsSync } from 'fs';

const __dirname = dirname(fileURLToPath(import.meta.url));
const PROJECT_ROOT = resolve(__dirname, '..');
const BIN = resolve(__dirname, 'bin/aristotle.js');
const TOPIC = process.argv[2] || 'coin flips';
const STDOUT_LOG = `/tmp/aristotle-tui-${Date.now()}.log`;
const JSON_LOG = STDOUT_LOG + '.chunks.jsonl';
const STATIC_GAP_MS = 1200; // gap longer than this = static window

writeFileSync(STDOUT_LOG, '');
writeFileSync(JSON_LOG, '');

const startedAt = Date.now();
function ts() { return ((Date.now() - startedAt) / 1000).toFixed(2).padStart(7) + 's'; }
function log(...a) { process.stderr.write(`[${ts()}] ` + a.map(x => typeof x === 'string' ? x : JSON.stringify(x)).join(' ') + '\n'); }

log(`topic: ${TOPIC}`);
log(`stdout log: ${STDOUT_LOG}`);
log(`chunk log:  ${JSON_LOG}`);

// Spawn aristotle under a PTY so Ink renders like a real terminal.
const term = pty.spawn(process.execPath, [BIN, TOPIC], {
  name: 'xterm-256color',
  cols: 120,
  rows: 40,
  cwd: '/tmp',
  env: { ...process.env, FORCE_COLOR: '1' },
});

let chunks = [];
let lastChunkAt = Date.now();
let sawDoneHint = false;
let sawError = false;

term.onData((data) => {
  const now = Date.now();
  const gap = now - lastChunkAt;
  const stripped = data.replace(/\x1b\[[0-9;?]*[A-Za-z]/g, '').replace(/\x1b[()][0-9A-Za-z]/g, '');
  chunks.push({ t: now - startedAt, gap, bytes: data.length, textLen: stripped.length, preview: stripped.slice(0, 120) });
  appendFileSync(STDOUT_LOG, data);
  appendFileSync(JSON_LOG, JSON.stringify({ t: now - startedAt, gap, bytes: data.length, preview: stripped.slice(0, 120) }) + '\n');
  lastChunkAt = now;
  if (/Book ready|open .*breakdown\.html|%%ARISTOTLE_DONE/i.test(stripped)) sawDoneHint = true;
  if (/error|exception/i.test(stripped)) sawError = true;
});

let exited = false;
term.onExit(({ exitCode, signal }) => {
  exited = true;
  log(`pty exited (code=${exitCode}, signal=${signal})`);
});

// Scripted input: wait for prompt, send, wait, repeat.
async function sleep(ms) { return new Promise(r => setTimeout(r, ms)); }
async function waitForText(needle, timeoutMs, description) {
  const deadline = Date.now() + timeoutMs;
  while (Date.now() < deadline) {
    if (exited) throw new Error(`pty exited while waiting for: ${description}`);
    // Scan all chunk previews for the needle (accumulate).
    const all = chunks.map(c => c.preview).join(' ');
    if (needle instanceof RegExp ? needle.test(all) : all.includes(needle)) return;
    await sleep(250);
  }
  throw new Error(`timeout (${timeoutMs}ms) waiting for: ${description}`);
}

function send(text) {
  log(`→ send: ${JSON.stringify(text)}`);
  term.write(text + '\r');
}

try {
  // Turn 1: topic is already the CLI arg. App auto-sends "I want to learn about: <topic>".
  // Wait for diagnosis questions to appear.
  await waitForText(/question|know|diagnos|calibrat/i, 45_000, 'diagnosis questions');
  log('  diagnosis rendered');

  // Turn 2: admit ignorance.
  await sleep(1500);
  send("I don't know any of the answers");

  // Wait for the outline + approval ask.
  await waitForText(/outline|chapter|approve|sound good|sound right/i, 90_000, 'outline with approval ask');
  log('  outline rendered');

  // Turn 3: approval. THIS is the moment the user complained about —
  // the 50s dead time after "ok".
  await sleep(1500);
  send('ok');

  // Wait for completion (or timeout). While waiting, keep capturing.
  const deadline = Date.now() + 15 * 60_000;
  while (Date.now() < deadline && !exited && !sawDoneHint) {
    await sleep(2000);
  }
  log(`completion: exited=${exited} sawDoneHint=${sawDoneHint}`);
} catch (err) {
  log(`FAIL during script: ${err.message}`);
} finally {
  await sleep(1000);
  if (!exited) term.kill();
  await sleep(500);
}

// --- Analyze ---
log('\n— chunk analysis —');
log(`total chunks: ${chunks.length}`);
const totalDur = chunks.length ? chunks[chunks.length - 1].t : 0;
log(`total duration: ${(totalDur / 1000).toFixed(1)}s`);

// Find gaps exceeding STATIC_GAP_MS.
const gaps = chunks
  .map((c, i) => ({ ...c, prev: i > 0 ? chunks[i - 1] : null }))
  .filter(c => c.gap > STATIC_GAP_MS);

log(`static windows (gap > ${STATIC_GAP_MS}ms): ${gaps.length}`);
gaps.slice(0, 20).forEach(g => {
  const at = (g.t / 1000).toFixed(2) + 's';
  log(`  @ ${at}  gap=${(g.gap / 1000).toFixed(2)}s  next="${g.preview.trim().slice(0, 80).replace(/\n/g, ' ⏎ ')}"`);
});

const longest = gaps.reduce((m, g) => g.gap > m.gap ? g : m, { gap: 0 });
log(`longest static window: ${(longest.gap / 1000).toFixed(2)}s @ ${(longest.t / 1000).toFixed(2)}s`);

// Pass/fail
const MAX_ALLOWED_GAP_S = 2.0;
const fail = gaps.filter(g => g.gap > MAX_ALLOWED_GAP_S * 1000);
if (fail.length === 0) {
  log(`\n  ✅ PASS — no static window > ${MAX_ALLOWED_GAP_S}s`);
  process.exit(0);
} else {
  log(`\n  🔴 FAIL — ${fail.length} windows exceeded ${MAX_ALLOWED_GAP_S}s of static display`);
  log(`  artifacts: ${STDOUT_LOG}, ${JSON_LOG}`);
  process.exit(2);
}
